---
title: "DoGE: Differential Gene Expression pipeline"
author: "Roberto Villegas-Diaz"
output:
  html_document:
    df_print: paged
urlcolor: blue
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE,warning=FALSE,echo=FALSE, cache = F)#,fig.height=5,fig.width=12)
options(knitr.table.format = function() {
  if (knitr::is_latex_output()) 
    "latex" else "pandoc"
})
options(knitr.kable.NA = '')
```

<img src="images/logo.jpg" style="width: 200px" />

```{r}
# Import libraries
library(GGally)
library(ggplot2)
library(gridExtra)
library(knitr)
library(kableExtra)
library(latex2exp)
library(pander)

library(foreach)
library(doParallel)
library(tictoc)
```

```{r}
## try http:// if https:// URLs are not supported
# BiocManager::install("edgeR") # Installing
library("edgeR") # Loading installed edgeR libraries
LFC <- 2

## Ensemble GTF can be extracted 
#gunzip Homo_sapiens.GRCh38.99.gtf.gz

# sed '/^[[:blank:]]*#/d;s/#.*//' Homo_sapiens.GRCh38.99.gtf | awk '($3 == "gene")' | awk -F";" '$1=$1' OFS="\t" > Homo_sapiens.GRCh38.99_gene_annotation.txt 
outpath <- paste0(getwd(),"/")

mydesign <- read.table("https://raw.githubusercontent.com/achalneupane/data/master/exp_design.csv", header=T, sep=",")
x <- read.table("6.COUNTS/counts.matrix", header=T, sep="\t", row.names=1)
# get rid of rows with zero values
x <- x[apply(x, 1, function(x) !all(x==0)),]

dbs <- read.table("https://raw.githubusercontent.com/achalneupane/data/master/Homo_sapiens.GRCh38.87_gene_annotation.txt", header=FALSE, sep="\t")
#head(dbs)
dbs <- dbs[,c(9, 11, 13)]
dbs <- sapply(dbs, function(x){gsub("gene_id|gene_name|gene_biotype| ", "", x)})
colnames(dbs) <- c("Geneid", "Gene", "Class")

kable(head(dbs), booktabs = TRUE) %>%
  kable_styling(latex_options = c("centered","striped","HOLD_position"))
#outpath <- "/media/owner/36bc7216-7321-48cc-b3f4-18f05a9fe11c/RNA-seq_bio792/"
```