---
# title: "DoGE: Differential Gene Expression pipeline
#         <img src='images/logo.jpg' style='width: 100px; float: right' />"
# author: "Roberto Villegas-Diaz"
pagetitle: "DoGE: Differential Gene Expression pipeline"
title: "<center>DoGE: Differential Gene Expression pipeline</center>"
author: "<center>See the code on [github](https://github.com/villegar/DoGE).</center>"
output:
  html_document:
    code_folding: "hide"
    css: style.css
    df_print: paged
    includes:
      before_body: header.html
      after_body: footer.html
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
urlcolor: blue
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE,warning=FALSE,echo=TRUE, cache = F,fig.height=6,fig.width=11, out.width="100%")
options(knitr.table.format = function() {
  if (knitr::is_latex_output())
    "latex" else "pandoc"
})
options(knitr.kable.NA = '')
```

<br><br>

<div class="mycontent">
```{r}
# Import libraries
library(ggplot2)
library(knitr)
library(kableExtra)

# library(foreach)
# library(doParallel)
# library(tictoc)

library(htmltools)
library(limma) # Bioconductor
library(DESeq2) # Bioconductor
library(dplyr)  # R package
library(latex2exp)
library(RColorBrewer)
library(gplots)
library(calibrate)
library(ggrepel)

baseWD <- "./" #paste0(getwd(),"/")
```

## MultiQC {.tabset .tabset-fade}
### FastQC
```{r, echo=FALSE}
#htmltools::includeHTML("Report_FastQC_Raw.html")
tags$iframe(
    title = "FastQC report",
    src = paste0(baseWD,"Report_FastQC_Raw.html"),
    width="100%",
    height="500",
    seamless="seamless",
    frameBorder="0")
```

### Trimming
```{r, echo=FALSE}
htmltools::tags$iframe(
    title = "Trimmed reads report",
    src = paste0(baseWD,"Report_Trimming.html"),
    width="100%",
    height="500",
    seamless="seamless",
    frameBorder="0")
```

### Trimming (FastQC)
```{r, echo=FALSE}
htmltools::tags$iframe(
    title = "FastQC for trimmed reads report",
    src = paste0(baseWD,"Report_FastQC_Trimmed.html"),
    width="100%",
    height="500",
    seamless="seamless",
    frameBorder="0")
```

### Alignment
```{r, echo=FALSE}
htmltools::tags$iframe(
    title = "Alignment report",
    src = paste0(baseWD,"Report_Alignment.html"),
    width="100%",
    height="500",
    seamless="seamless",
    frameBorder="0")
```

## DESeq2 Analysis
### 1. Reading data

First we read in counts data inside the `counts.matrix` file, the experimental design (`exp_design.csv`), and the gene annotation table (`gene_annotation.txt`).
```{r, cache=TRUE}
exp_design <- read.table(paste0(baseWD,"exp_design.csv"), header=T, sep=",")
countData <- read.table(paste0(baseWD,"counts.matrix"), header=T, sep="\t", row.names=1)

# get rid of rows with zero values
countData <- countData[apply(countData, 1, function(countData) !all(countData==0)),]

# Load annotation table
dbs <- read.table(paste0(baseWD,"gene_annotation.txt"), header=FALSE, sep="\t")

dbs <- dbs[,c(9, 11, 13)]
dbs <- sapply(dbs, function(x){gsub("gene_id|gene_name|gene_biotype| ", "", x)})
colnames(dbs) <- c("Geneid", "Gene", "Class")
dbs <- data.frame(dbs)
```


```{r read_counts, echo=FALSE}
kable(head(countData, 5), booktabs = TRUE, caption = "Top 5 genes read counts") %>%
  kable_styling(latex_options = c("centered","striped","HOLD_position","scale_down"))
```

```{r exp_design, echo=FALSE}
kable(exp_design, booktabs = TRUE, caption = "Experimental design") %>%
  kable_styling(latex_options = c("centered","striped","HOLD_position","scale_down"))
```

```{r dbs, echo=FALSE}
kable(head(dbs, 5), booktabs = TRUE, caption = "Top 5 genes in the annotation table") %>%
  kable_styling(latex_options = c("centered","striped","HOLD_position","scale_down"))
```


### 2. Simple exploration of the raw data

```{r, results='asis', echo=FALSE}
summary2kable <- function(summary, caption){
  tmp <- strsplit2(as.data.frame(summary)$Freq,":")
  summary_cols <- unique(tmp[,1])
  tmp <- matrix(as.numeric(tmp[,2]),ncol = length(summary_cols), byrow = T)
  aux <- data.frame(tmp)
  rownames(aux) <- colnames(summary)
  colnames(aux) <- summary_cols
  cat(kable(aux, booktabs = TRUE, caption = caption) %>%
    kable_styling(latex_options = c("centered","striped","HOLD_position","scale_down")))
}

summary2kable(summary(countData), caption = "Summary for read counts")
```


```{r barplot-cpm}
barplot(colSums(countData)/1e6, col="#66BD63",las=1,main="Total read counts (millions)", ylab="Total reads in millions")
```

```{r hist-log2-lib1}
transformation <- paste0("Log_2(",colnames(countData)[1],")")
hist(log2(countData[,1]),br=200, xlab = TeX(transformation), col="#08519C",
     main = TeX(paste0("Histogram of ",transformation)))
```

#### Simple log transformation on unnormalized data
```{r}
logCountData = log2(1 + countData)
par(mfrow = c(1, 2), mar = c(8,2,2,2))  # two columns
hist(logCountData[,1], xlab = TeX(transformation),
     main = TeX(paste0("Histogram of ",transformation)))
boxplot(logCountData,las=3)
```

```{r}
x <- logCountData
myColors <- brewer.pal(dim(x)[2],"Spectral")#rainbow(dim(x)[2])
plot(density(x[,1]),col = myColors[1], lwd=2,
  xlab="Expression values", ylab="Density", main= "Distribution of transformed data",
  ylim=c(0, max(density(x[,1])$y)+.02 ) )

for(i in 2:dim(x)[2] )
lines(density(x[,i]),col=myColors[i], lwd=2 )
  legend("topright", cex=1.1,colnames(x), lty=rep(1,dim(x)[2]), col=myColors)
```

```{r}
xlab <- paste0("Log_2(",colnames(countData)[1],")")
ylab <- paste0("Log_2(",colnames(countData)[2],")")
plot(logCountData[,1],logCountData[,2],
     xlab = TeX(xlab),
     ylab = TeX(ylab),
     main = TeX(paste0(ylab, " vs ",xlab)))
```

### 3. Filtering, normalization, and transformation using DESeq2

Using the `condition` column from the experimental design

```{r}
condition <- exp_design$condition
countdata <- as.matrix(countData)
coldata <- data.frame(row.names=colnames(countdata), condition)
dds <- DESeqDataSetFromMatrix(countData=countdata, colData=coldata, design=~condition)

# Run the DESeq pipeline
dds <- DESeq(dds)

# Plot dispersions
plotDispEsts(dds, main="")
```

<!-- ### Filtering  -->
```{r, eval=FALSE}
dds <- dds[ rowSums(counts(dds)) > 5, ]
```

#### rlog transformation
Regularized log transformation
```{r}
# Regularized log transformation for clustering/heatmaps, etc
rld <- rlog(dds, blind = FALSE)
kable(head(assay(rld), 5), booktabs = TRUE) %>%
  kable_styling(latex_options = c("centered","striped","HOLD_position","scale_down"))
```

#### VSD transformation
Variance-stabilizing transformation
```{r}
vsd <- vst(dds, blind = FALSE)

kable(head(assay(vsd), 5), booktabs = TRUE) %>%
  kable_styling(latex_options = c("centered","striped","HOLD_position","scale_down"))
```

```{r}
dds <- estimateSizeFactors(dds)
kable(t(sizeFactors(dds)), booktabs = TRUE) %>%
  kable_styling(latex_options = c("centered","striped","HOLD_position","scale_down"))
```

#### Started log on scaled data
Usings the normalized=TRUE option in the `counts()` method of `DESeq2`, we adjust for different library sizes.
```{r}
slog <- log2(counts(dds, normalized=TRUE)+1)
kable(head(slog, 5), booktabs = TRUE) %>%
  kable_styling(latex_options = c("centered","striped","HOLD_position","scale_down"))
```

```{r, eval=F}
par(mfrow = c(1, 3))  # 3 columns
plot(slog[,1],slog[,2])
plot(assay(rld)[,1],assay(rld)[,2])
plot(assay(vsd)[,1],assay(vsd)[,2])
```

```{r}
df <- bind_rows(
  as_tibble(slog[,1:2]) %>%
         mutate(transformation = "log2(x + 1)"),
  as_tibble(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"),
  as_tibble(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"))

colnames(df)[1:2] <- c("x", "y")  

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation)
```

### 4. Exploratory Data Analysis
#### Sample distance heatmap
```{r, fig.align='center', fig.width=6}
mycols <- brewer.pal(length(unique(condition)), "YlGnBu")[1:length(unique(condition))]
mycols[1] <- "#D7191C"

# Sample distance heatmap
sampleDists <- as.matrix(dist(t(assay(rld))))

heatmap.2(as.matrix(sampleDists), key=F, trace="none",
          col=colorpanel(100, "black", "white"),
          ColSideColors=mycols[condition], RowSideColors=mycols[condition],
          margin=c(10, 10), main="Sample Distance Matrix")
```

#### Principal Component Analysis (PCA)    
```{r, fig.align='center', fig.width=7}
pcaData <- plotPCA(rld, intgroup = c("condition"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=condition)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  coord_fixed() +
  geom_text_repel(aes(PC1, PC2, label = rownames(pcaData)), pcaData) +
  scale_color_manual(name="Condition", values = mycols) +
  theme(aspect.ratio=1)
```

#### K-means clustering of genes

```{r kmeans, fig.height=10}
k <- 5 # Number of clusters
n <- 100 # number of top genes by standard deviation
x <- assay(rld)
if(n>dim(x)[1]) n = dim(x)[1] # max	as data
x <- x[order(apply(x,1,sd),decreasing=TRUE),]  # sort genes by standard deviation
x <- x[1:n,]   # only keep the n genes

x <- 100* x[1:n,] / apply(x[1:n,],1,function(y) sum(abs(y))) # L1 norm

set.seed(2)
#k = input$k  # number of clusters

cl <- kmeans(x,k,iter.max = 50)

hc <- hclust(dist(cl$centers-apply(cl$centers,1,mean) )  )
tem <- match(cl$cluster,hc$order) #  new order
x <- x[order(tem),]
bar <- sort(tem)

# heatmap with color bar define gene groups
myheatmap2 <- function (x,bar,n=-1) {
	# number of genes to show
  ngenes = as.character(table(bar))
	if(length(bar) >n && n != -1) {ix = sort( sample(1:length(bar),n) ); bar = bar[ix]; x = x[ix,]}

	# this will cutoff very large values, which could skew the color
  gene_names <- rownames(x)
	x <- as.matrix(x)-apply(x,1,mean)

	cutoff <- median(unlist(x)) + 3*sd (unlist(x))
	x[x>cutoff] <- cutoff
	cutoff <- median(unlist(x)) - 3*sd (unlist(x))
	x[x< cutoff] <- cutoff
	#colnames(x)= detectGroups(colnames(x))
  #row.names(x) <- gene_names
	heatmap.2(x,  Rowv =F,Colv=F, dendrogram ="none",
	          col=colorpanel(75, "black", "white"),
	          #col=greenred(75),
	          labRow = gene_names,
	          cexRow=0.6,
	          density.info="none",
	          trace="none", scale="none", keysize=.3,
	          key=F,RowSideColors = mycolors[bar],
	          margins = c(8, 24), srtCol=45
	)
	legend.text = paste("Cluster ", toupper(letters)[unique(bar)], " (N=", ngenes,")", sep="")
	par(lend = 1)           # square line ends for the color legend
	legend("topright",      # location of the legend on the heatmap plot
		legend = legend.text, # category labels
		col = mycolors,  # color key
		lty= 1,             # line style
		lwd = 10            # line width
	)
}

mycolors <- brewer.pal(k,"YlGnBu")
mycolors[1] <- "#D7191C"
myheatmap2(x-apply(x,1,mean), bar,1000)
```
<br><br><br>
</div>
